flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/ExecutionContext.java

```diff
@@ -0,0 +1,219 @@
+ * multiple queries as long as the session context does not change. This must be thread-safe as it
+ * might be reused across different query submissions.
+ */
+public class ExecutionContext {
+
+    // TODO: merge the ExecutionContext into the SessionContext.
+    // Members that should be reused in the same session.
+    private final Environment environment;
+    private final Configuration flinkConfig;
+    private final SessionState sessionState;
+    private final URLClassLoader classLoader;
+
+    private final TableEnvironment tableEnv;
+
+    public ExecutionContext(
+            Environment environment,
+            Configuration flinkConfig,
+            URLClassLoader classLoader,
+            SessionState sessionState) {
+        this.environment = environment;
+        this.flinkConfig = flinkConfig;
+        this.sessionState = sessionState;
+        this.classLoader = classLoader;
+
+        this.tableEnv = createTableEnvironment();
+    }
+
+    /**
+     * Create a new {@link ExecutionContext}.
+     *
+     * <p>It just copies from the {@link ExecutionContext} and rebuild a new {@link
+     * TableEnvironment}. But it still needs {@link Environment} because {@link Environment} doesn't
+     * allow modification.
+     *
+     * <p>When FLINK-21462 finishes, the constructor only uses {@link ExecutionContext} as input.
+     */
+    public ExecutionContext(Environment environment, ExecutionContext context) {
+        this.environment = environment;
+        this.flinkConfig = context.flinkConfig;
+        this.sessionState = context.sessionState;
+        this.classLoader = context.classLoader;
+        // create a new table env
+        this.tableEnv = createTableEnvironment();
+    }
+
+    /**
+     * Executes the given supplier using the execution context's classloader as thread classloader.
+     */
+    public <R> R wrapClassLoader(Supplier<R> supplier) {
+        try (TemporaryClassLoaderContext ignored = TemporaryClassLoaderContext.of(classLoader)) {
+            return supplier.get();
+        }
+    }
+
+    public TableEnvironment getTableEnvironment() {
+        return tableEnv;
+    }
+
+    // ------------------------------------------------------------------------------------------------------------------
+    // Helper to create Table Environment
+    // ------------------------------------------------------------------------------------------------------------------
+
+    private TableEnvironment createTableEnvironment() {
+        EnvironmentSettings settings = environment.getExecution().getEnvironmentSettings();
+        TableConfig config = new TableConfig();
+        config.addConfiguration(flinkConfig);
+        // Override the value in configuration.
+        // TODO: use `table.planner` and `execution.runtime-mode` to configure the TableEnvironment
+        config.addConfiguration(settings.toConfiguration());
+
+        if (environment.getExecution().isStreamingPlanner()) {
```

Comment:
It seems the new introduced `table.planner` and `execution.mode` option doesn't take effect here. Currently, the implementation only depends on the SQL Client configuration. 

I think we should use `table.planner` and `execution.mode`  to decide how to create TableEnv. 

Besides, please add tests for this.

Reply:
Sure. But I think we can leave the work to FLINK-21462. It will do the work about Configuration and Yaml compatibility.